controllers/topology/internal/scope/blueprint.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+// ClusterBlueprint holds all the objects required for computing the desired state of a managed Cluster topology,
+// including the ClusterClass and all the referenced templates.
+type ClusterBlueprint struct {
```

Comment:
Could we plan to add some additional documentation about these types? There's  semantic overlap between Template and Blueprint that's confusing at first glance at least. The flow here of Scope -> ClusterBlueprint -> ControlPlaneBlueprint -> Template etc. is deep and could be hard to follow

Reply:
The questions 'What is a blueprint?' and 'What is a scope' need good answers, especially if this pattern is repeated in future.

Reply:
> Could we plan to add some additional documentation about these types? 

Blueprint was introduced to replace class, which in turns was creating confusion with ClusterClass.

Current definitions are:
```
// ClusterBlueprint holds all the objects required for computing the desired state of a managed Cluster topology,
// including the ClusterClass and all the referenced templates.
type ClusterBlueprint struct {
   ...
}

// Scope holds all the information to process a request in the topology/ClusterReconciler controller.
type Scope struct {
   ...
}
```

Happy to improve them if there are suggestions...

> The flow here of Scope -> ClusterBlueprint -> ControlPlaneBlueprint -> Template etc. is deep and could be hard to follow

Nested types are hard to follow, but on the other side they provide an organization in this struct, which reads as
`s.Blueprint.controlplane.template` when used in the code.

> The questions 'What is a blueprint?' and 'What is a scope' need good answers, especially if this pattern is repeated in future.

I'm not sure the blueprint is something that will become a pattern, topology/ClusterRecociler is quite different from existing reconcilers.

Scope instead is already a pattern in some providers (e.g. CAPA); also KCP has a complicated variant of it; the name is kind of debatable, given that it is a request context, but for this first iteration I stuck to what is used in the community

Reply:
With that context scope definitely makes sense - It would be good to get a one-line answer to differentiate what a Blueprint is versus a Template (or even a Spec!) though, but it's not super important unless the patter might be repeated elsewhere.